Job queue

pstibrany opened the pull request:
This PR reimplements `chan chunkWriteJob` with custom buffered queue that should use less memory, because it doesn't preallocate entire buffer for maximum queue size at once. Instead it allocates individual "segments" with smaller size.

As elements are added to the queue, they fill individual segments. When elements are removed from the queue (and segments), empty segments can be thrown away. This doesn't change memory usage of the queue when it's full, but should decrease its memory footprint when it's empty (queue will keep max 1 segment in such case).

(Going to test the change in our dev cluster)

replay commented on tsdb/chunks/queue.go:
```diff
@@ -0,0 +1,132 @@
+	mtx         sync.Mutex            // protects all following variables
+	pushed      *sync.Cond            // signalled when something is pushed into the queue
+	popped      *sync.Cond            // signalled when element is popped from the queue
+	first, last *writeJobQueueSegment // pointer to first and last segment, if any
```
that's interesting, i haven't seen this syntax before where you define to properties in one line like this. It makes sense here, but if we use that here why not also for `pushed` and `popped`?

pstibrany replied:
Sure, let's do that. I will merge their comments.

replay commented on tsdb/chunks/queue.go:
```diff
@@ -0,0 +1,132 @@
+	q.mtx.Lock()
+	defer q.mtx.Unlock()
+
+	for {
+		if q.closed {
+			return false
+		}
+
+		if q.size >= q.maxSize {
+			// wait until queue has more space or is closed
+			q.popped.Wait()
+			continue
+		}
+
+		// cap(q.last.queue)-len(q.last.queue) is free space remaining in the q.last.queue.
+		if q.last == nil || cap(q.last.queue)-len(q.last.queue) == 0 {
+			prevLast := q.last
+			q.last = &writeJobQueueSegment{
+				queue: make([]chunkWriteJob, 0, q.segmentSize),
+			}
+
+			if prevLast != nil {
+				prevLast.next = q.last
+			}
+			if q.first == nil {
+				q.first = q.last
+			}
+		}
+
+		q.last.queue = append(q.last.queue, job)
+		q.size++
+		q.pushed.Signal()
+		return true
+	}
```
Does all of this need to be inside the loop, wouldn't something like this be sufficient:

```suggestion
	if q.closed {
		return false
	}

	for {
		if q.size >= q.maxSize {
			// wait until queue has more space or is closed
			q.popped.Wait()
			continue
		}
	}

	// cap(q.last.queue)-len(q.last.queue) is free space remaining in the q.last.queue.
	if q.last == nil || cap(q.last.queue)-len(q.last.queue) == 0 {
		prevLast := q.last
		q.last = &writeJobQueueSegment{
			queue: make([]chunkWriteJob, 0, q.segmentSize),
		}

		if prevLast != nil {
			prevLast.next = q.last
		}
		if q.first == nil {
			q.first = q.last
		}
	}

	q.last.queue = append(q.last.queue, job)
	q.size++
	q.pushed.Signal()
	return true
```

pstibrany replied:
You're right!

pstibrany commented:
@replay Thanks for your review. I've addressed your comments. I've also modified implementation to avoid using `len` and `cap` and use two indexes instead (index of next read for `pop`, and next write for `push`), I think it makes code easier to read and understand.

I would appreciate if you had a chance to look at the PR again.

replay commented on tsdb/chunks/queue.go:
```diff
@@ -0,0 +1,132 @@
+
+		if q.size >= q.maxSize {
+			// wait until queue has more space or is closed
+			q.popped.Wait()
```
Oh right, I need to read the docs of `cond.Wait` again 🤦‍♂️

pstibrany commented on tsdb/chunks/queue.go:
```diff
@@ -0,0 +1,132 @@
+}
+
+func newWriteJobQueue(maxSize, segmentSize int) *writeJobQueue {
+	if maxSize <= 0 || segmentSize <= 0 {
```
```
┌────────────────────────────────────────────┐
│                                            │
│                  segment                   │
│                                            │
└────────────────────────────────────────────┘
      ▲                              ▲        
      │                              │        
      │                              │        
┌──────────┐                   ┌──────────┐   
│ nextRead │                   │nextWrite │   
└──────────┘                   └──────────┘   
```

Let's imagine that both `nextRead` and `nextWrite` move towards right. Adding new element to the queue = moving `nextWrite` by one element. Popping element from the queue = moving `nextRead` by one element. Once `nextRead` reaches the end, we drop the segment.

When size == 0, `nextRead` will be equal to `nextWrite`, but another `push` will use another element (`nextWrite`).

If queue is still used (pushed to / popped from), eventually segment will be used fully, even if it's much bigger than `maxSize`.

replay commented on tsdb/chunks/queue.go:
```diff
@@ -0,0 +1,126 @@
+	defer q.mtx.Unlock()
+
+	// wait until something is pushed to the queue, or queue is closed.
+	for q.size == 0 {
+		if q.closed {
+			return chunkWriteJob{}, false
+		}
+
+		q.pushed.Wait()
+	}
```
Do I understand it correctly that if there are elements in the queue and then the queue gets closed, these elements that are already in it can still be popped until the queue is empty?

What happens if `pop()` is called on an empty queue and blocks on `q.pushed.Wait()`, but then the queue gets closed and `q.pushed.Wait()` unblocks? Wouldn't that panic on `:105` because `q.first` would be `nil`?

replay replied:
Ah right, never mind me :) thanks!

replay reviewed (approved):
This looks great, thanks!

pstibrany commented:
Thanks for your review @replay!

pracucci commented on tsdb/chunks/chunk_write_queue.go:
```diff
@@ -230,7 +251,7 @@ func (c *chunkWriteQueue) queueIsEmpty() bool {
 func (c *chunkWriteQueue) queueIsFull() bool {
 	// When the queue is full and blocked on the writer the chunkRefMap has one more job than the cap of the jobCh
 	// because one job is currently being processed and blocked in the writer.
```
This comment should be updated too.

pracucci replied:
I think comment in `queueSize()` needs to be updated too.

pstibrany replied:
Oh, thanks for catching this. Will send new PR.
